internal/addonsdk/sdk.go

```diff
@@ -0,0 +1,192 @@
+func SetupStatusReporter(addonInstanceInteractor client, addonName string, addonTargetNamespace string, logger logr.Logger) StatusReporterClient {
+	initializeSingletonOnce.Do(func() {
+		statusReporterSingleton = &statusReporter{
+			addonInstanceInteractor: addonInstanceInteractor,
+			addonName:               addonName,
+			addonTargetNamespace:    addonTargetNamespace,
+			latestConditions: []metav1.Condition{
+				{
+					Type:    AddonHealthyConditionType,
+					Status:  metav1.ConditionUnknown,
+					Reason:  "NoHealthReported",
+					Message: fmt.Sprintf("Addon %q hasn't reported health yet", addonName),
+				},
+			},
+			stopperCh: make(chan bool),
+			updateCh:  make(chan updateOptions),
+			doneCh:    make(chan bool),
+			log:       logger,
+		}
+		// because the heartbeat reporter still hasn't been started
+		defer close(statusReporterSingleton.doneCh)
+	})
+	return statusReporterSingleton
+}
+
+func (sr *statusReporter) Start(ctx context.Context) error {
+	// ensures to tie only one heartbeat-reporter loop at a time to a StatusReporter object
+	select {
+	case <-sr.doneCh:
+		sr.log.Info("StatusReporter already found to be running. Please explicitly Stop() it first before trying to restart it")
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+	sr.doneCh = make(chan bool)
+	defer close(sr.doneCh)
+
+	currentAddonInstance := &addonsv1alpha1.AddonInstance{}
+	if err := sr.addonInstanceInteractor.GetAddonInstance(ctx, types.NamespacedName{Name: "addon-instance", Namespace: sr.addonTargetNamespace}, currentAddonInstance); err != nil {
+		return fmt.Errorf("error occurred while fetching the current heartbeat update period interval: %w", err)
+	}
+	sr.interval = currentAddonInstance.Spec.HeartbeatUpdatePeriod.Duration
+	defer sr.ticker.Stop()
+	sr.ticker = time.NewTicker(sr.interval)
+
+	for {
+		select {
+		case update := <-sr.updateCh:
+			// update the interval if the newInterval in the `update` is provided and is not equal to the existing interval
+			// synchronize the timer with this new interval
+			if update.addonInstance != nil {
+				if update.addonInstance.Spec.HeartbeatUpdatePeriod.Duration != sr.interval {
+					sr.interval = update.addonInstance.Spec.HeartbeatUpdatePeriod.Duration
+					sr.ticker.Reset(sr.interval)
+				}
+			}
+
+			if update.conditions != nil {
+				sr.latestConditions = *update.conditions
+			}
+		case <-sr.ticker.C:
+			if err := sr.updateAddonInstanceStatus(ctx, sr.latestConditions); err != nil {
+				sr.log.Error(err, "failed to report the regular heartbeat")
+			}
+		case <-ctx.Done():
+			return nil
+		case <-sr.stopperCh:
+			return nil
+		}
+	}
+}
+
+func (sr *statusReporter) Stop(ctx context.Context) error {
+	select {
+	case <-sr.doneCh: // will non-blockingly receive whenever doneCh would be closed
+		sr.log.Info("status reporter is already stopped")
+		return nil
+	case sr.stopperCh <- true:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
```

Comment:
This method is redundant with the canceling the context handed to `Start()`:
e.g. these two code blocks are the same:
```go
func do() {
  ctx, cancel := context.WithCancel(context.Background())
  defer cancel()
  go statusReporter.Start(ctx)
}
```

```go
func do() {
  defer statusReporter.Stop(context.Background())
  go statusReporter.Start(context.Background())
}
```

Reply:
Ahh I see, nice one. With that, we can even get rid of the `stopperCh`